### PR TITLE
[HOLD] Add "Mark as Done" button

### PR DESCRIPTION
**Pull request by Gonals**
### Fixed Issues

$ https://github.com/Expensify/App/issues/30258
PROPOSAL: 

### Tests
1. You'll need two accounts for this, so let's start by creating them. A `user` account and an `admin` account.
2. Log into newdot from the admin account and create a new workspace. Invite the user account to it.
3. Log into into newdot from the user account and request money (manually) from the workspace.
Screenshot 2023-11-07 at 5 57 35 PM

4. Log into OldDot from the admin account and go to `Reports`. You should see the request report:
Screenshot 2023-11-07 at 5 57 51 PM

5. After a couple of seconds, log into newdot from the admin account and open the workspace chat with the user account. The request should now have a "Mark as done" button (you might need to refresh or open the request to see the right button. This PR does not cover pushing the reimbursable info to Onyx after it changes in OldDot):
Screenshot 2023-11-07 at 6 00 33 PM

6. Click it. Confirm the report gets closed, but is still accessible in NewDot. The request should display like so:
Screenshot 2023-11-07 at 6 01 37 PM
Screenshot 2023-11-07 at 6 01 45 PM
 

- [ ] Verify that no errors appear in the JS console

### QA Steps

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
        - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

Screenshot 2023-11-09 at 2 20 32 PM
Screenshot 2023-11-09 at 2 21 44 PM

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

Screenshot 2023-11-09 at 1 40 50 PM

Screenshot 2023-11-09 at 1 45 54 PM

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

Screenshot 2023-11-08 at 4 32 36 PM

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Comment by Gonals** [comment]
Mobile web screenshot missing until we figure out how to run mobile web with the new address 😅

**Comment by Gonals** [comment]
Ugh. I hate typescript. Can anyone tell my why it is complaining? `OriginalMessage.ts` seems to include `IOUReportID`

File: src/libs/ReportActionsUtils.ts [hunk 1]
```diff
@@ -314,7 +318,16 @@ function shouldReportActionBeVisible(reportAction: OnyxEntry<ReportAction>, key:
 
     // Ignore closed action here since we're already displaying a footer that explains why the report was closed
     if (reportAction.actionName === CONST.REPORT.ACTIONS.TYPE.CLOSED) {
-        return false;
+        if (!allReports) {
+            return false;
+        }
+        const expenseReport = allReports[reportAction?.originalMessage?.IOUReportID];
+        return (
+            expenseReport !== null &&
```
Comment (marcaaron) [inline src/libs/ReportActionsUtils.ts:326]: Kinda looks like maybe the reportAction is not picking up the types from here properly...

https://github.com/Expensify/App/blob/dee9b3110a7ebe0aa250c3f966d6a69925d916f4/src/types/onyx/ReportAction.ts#L109-L109

That's an "intersection type" so it should merge those two together...

And if you look at `OriginalMessageIOU` it should include `IOUReportID`

https://github.com/Expensify/App/blob/dee9b3110a7ebe0aa250c3f966d6a69925d916f4/src/types/onyx/OriginalMessage.ts#L18-L36

So, it appears that the union types from here: 

https://github.com/Expensify/App/blob/dee9b3110a7ebe0aa250c3f966d6a69925d916f4/src/types/onyx/OriginalMessage.ts#L181-L195

Are overwriting each other when the intersection type gets "merged". The one it thinks you have is this:

https://github.com/Expensify/App/blob/dee9b3110a7ebe0aa250c3f966d6a69925d916f4/src/types/onyx/OriginalMessage.ts#L90-L97

Comment (Gonals) [reply src/libs/ReportActionsUtils.ts:326]: Huh. I tried creating a `OriginalMessageDone` with my structure and didn't help, probably because the actionName is the same. I'll try editing that one and seeing if it likes that better

Comment (Gonals) [reply src/libs/ReportActionsUtils.ts:327]: Yep, that did it

File: src/libs/ReportActionsUtils.ts [hunk 2]
```diff
@@ -80,6 +80,10 @@ function isMoneyRequestAction(reportAction: OnyxEntry<ReportAction>): boolean {
     return reportAction?.actionName === CONST.REPORT.ACTIONS.TYPE.IOU;
 }
 
+function isDoneAction(reportAction: OnyxEntry<ReportAction>): boolean {
+    return reportAction?.actionName === CONST.REPORT.ACTIONS.TYPE.CLOSED;
```
Comment (marcaaron) [inline src/libs/ReportActionsUtils.ts:84]: Full disclosure I did not read the design doc related to this change. The CLOSED action also used to indicate an "archived" chat. So, maybe we should also check the report to see what kind it is? I am lacking the full context though about why we decided to reuse this action to mean both "done" or "archived".

File: src/libs/ReportActionsUtils.ts [hunk 3]
```diff
@@ -324,7 +328,16 @@ function shouldReportActionBeVisible(reportAction: OnyxEntry<ReportAction>, key:
 
     // Ignore closed action here since we're already displaying a footer that explains why the report was closed
     if (reportAction.actionName === CONST.REPORT.ACTIONS.TYPE.CLOSED) {
-        return false;
+        if (!allReports) {
+            return false;
+        }
```
Comment (marcaaron) [inline src/libs/ReportActionsUtils.ts:333]: uh oh - why would this not exist?

File: src/libs/ReportActionsUtils.ts [hunk 4]
```diff
@@ -324,7 +328,16 @@ function shouldReportActionBeVisible(reportAction: OnyxEntry<ReportAction>, key:
...
+        const expenseReport = allReports[reportAction?.originalMessage?.IOUReportID ?? 0] ?? null;
```
Comment (marcaaron) [inline src/libs/ReportActionsUtils.ts:334]: A bit hard to parse with all the `?`  in it.

Going to take a leap and say that if we have no IOUReportID at this stage then something went wrong. And another leap that if the expense report doesn't exist then another thing went wrong.

File: src/libs/ReportUtils.js [hunk 5]
```diff
@@ -638,7 +638,7 @@ function findLastAccessedReport(reports, ignoreDomainRooms, policies, isFirstTim
  * @returns {Boolean}
  */
 function isArchivedRoom(report) {
-    return report && report.statusNum === CONST.REPORT.STATUS.CLOSED && report.stateNum === CONST.REPORT.STATE_NUM.SUBMITTED;
+    return report && report.statusNum === CONST.REPORT.STATUS.CLOSED && report.stateNum === CONST.REPORT.STATE_NUM.SUBMITTED && report.type !== CONST.REPORT.TYPE.EXPENSE;
```
Comment (marcaaron) [inline src/libs/ReportUtils.js:641]: Ok, haha, it's making somewhat more sense now. We really repurposed the "CLOSE" action for chats and now we're in a weird spot where we need to check to see if this is a chat...

Small thing - but let's check if `report.type == 'chat'` instead of "not expense report"?

File: src/libs/ReportUtils.js [hunk 6]
```diff
@@ -1485,6 +1485,21 @@ function hasNonReimbursableTransactions(iouReportID) {
...
+ *
+ * @param {String} iouReportID
+ * @returns {Boolean}
+ */
+function hasReimbursableTransactions(iouReportID) {
+    const allTransactions = TransactionUtils.getAllReportTransactions(iouReportID);
+    return _.filter(allTransactions, (transaction) => transaction.reimbursable === true).length > 0;
+}
+
+function isMarkedAsDone(report) {
```
Comment (marcaaron) [inline src/libs/ReportUtils.js:1499]: JS doc here too?

File: src/libs/ReportUtils.js [hunk 7]
```diff
@@ -2745,6 +2765,44 @@ function buildOptimisticSubmittedReportAction(amount, currency, expenseReportID)
...
+            {
+                style: 'strong',
+                text: lodashGet(currentUserPersonalDetails, 'displayName', currentUserEmail),
+                type: 'TEXT',
+            },
+        ],
+        reportActionID: NumberUtils.rand64(),
+        shouldShow: true,
+        created: DateUtils.getDBTime(),
+        pendingAction: CONST.RED_BRICK_ROAD_PENDING_ACTION.ADD,
```
Comment (marcaaron) [inline src/libs/ReportUtils.js:2802]: **noting:** More than a few of these properties could be moved into a single method that returns some kind of "base" report action properties and reduce the repetition we have in other methods.

File: src/libs/actions/IOU.js [hunk 8]
```diff
@@ -2713,6 +2713,76 @@ function submitReport(expenseReport) {
     );
 }
 
+function markAsDone(chatReport, iouReport) {
```
Comment (marcaaron) [inline src/libs/actions/IOU.js:2716]: JS doc here?

**Review by marcaaron — COMMENTED** [review]
Changes look pretty good! 

I found it slightly confusing that we are using the same report action (`CLOSE`) to indicate a "done report" or "archived chat". But don't really have a clear suggestion other than go back in time and create a new "archived" report action for the chats.

**Comment by Gonals** [comment]
Putting this back on HOLD, as we may be removing/changing this functionality

**Comment by marcaaron** [comment]
Not sure what's up with this PR so just gonna peel myself off of the review for now in hopes that it will drop off my K2.

**Comment by marcaaron** [comment]
Oh that's not a thing heh. @Gonals can we close this one ?

**Comment by Gonals** [comment]
> Oh that's not a thing heh. @Gonals can we close this one ?

Yep, I think so
